ts/keys.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Matt Tracy (matt.r.tracy@gmail.com)
+
+package ts
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/proto"
+	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util/encoding"
+)
+
+// Time series keys are carefully constructed to usefully sort the data in the
+// key-value store for the purpose of queries. Time series data is queryied by
+// providing a series name and a timespan; we therefore expose the series name
+// and the timestamp prominently on the key.
+//
+// The precise formula for a time series key is:
+//   [system key prefix]tsd[series name][resolution][timestamp][source key]
+//
+// The series name is an arbitrary string identifying the series, although the
+// ts system may enforce naming rules at a higher level. This string is binary
+// encoded in the key.
+//
+// The resolution refers to the sample duration at which data is stored.
+// Cockroach supports a fixed set of named resolutions, which are stored in the
+// Resolution enumeration. This value is encoded as a VarInt in the key.
+//
+// Each time series data key contains the data for a specific range of time. The
+// timestamp value in the key is the beginning of that range; in other words,
+// the earliest time for which this key might contain data. Cockroach divides
+// all data for a series into slots of uniform length based on the "key
+// duration" of the Resolution.
+//
+// The source key is a possibly empty string which identifies the source from
+// which the series data was gathered. Data for a series may be gathered from
+// multiple sources, which are stored separately but are sorted adjacently for
+// efficient access.
+var (
+	// keyDataPrefix is the key prefix for time series data keys.
+	keyDataPrefix = proto.MakeKey(engine.KeySystemPrefix, proto.Key("tsd"))
+)
+
+// MakeDataKey creates a time series data key for the given series name,
+// source, resolution and timestamp.
+func MakeDataKey(name string, source string, r Resolution,
+	timestamp int64) proto.Key {
+	return proto.MakeKey(keyDataPrefix, encoding.EncodeBytes(nil, []byte(name)),
+		encoding.EncodeVarint(nil, int64(r)), encoding.EncodeVarint(nil, timestamp),
+		[]byte(source))
```

Comment:
You really should be building up the key incrementally. That is why `encoding.EncodeBytes` takes a slice to append to. Something like:

```
  k := append([]byte(nil), keyDataPrefix...)
  k = encoding.EncodeBytes(k, []byte(name))
  k = encoding.EncodeVarint(k, int64(r))
  k = encoding.EncodeVarint(k, timestamp)
  k = append(k, source...)
  return proto.Key(k)
```

Does a particular resolution imply a particular key duration? If yes, should `MakeDataKey` be normalizing the timestamp? For example, `timestamp /= r.KeyDuration()`.